feat: add a notifier to the timerService

Chris-Hibbert opened the pull request:
includes tests.

renames `createRepeater()` to `makeRepeater()`, and adds `makeNotifier()`.
moves type declarations for TimerService to SwingSet (from zoe).
add makeNotifier to the manualTimer.

closes: #2002
closes: https://github.com/Agoric/dapp-token-economy/issues/32

erights commented on packages/SwingSet/src/vats/vat-timerWrapper.js:
```diff
@@ -41,6 +42,29 @@ export function buildRootObject(vatPowers) {
+          Nat(delaySecs) >= 0,
+          details`createNotifier's first parameter must be a non-negative integer. ${delaySecs}`,
+        );
+        assert(
+          Nat(interval),
+          details`createNotifier's second parameter must be an integer, ${interval}`,
+        );
```
The test and the error text disagree. I see the same issue on line 26 above. A negative integer will be rejected with a complaint that it is not an integer.

erights commented on packages/SwingSet/src/vats/vat-timerWrapper.js:
```diff
@@ -41,6 +42,29 @@ export function buildRootObject(vatPowers) {
         repeaters.set(index, vatRepeater);
         return vatRepeater;
       },
+      createNotifier(delaySecs, interval) {
```
Is the type of this declared somewhere? Is there a doc-comment explaining what it does? Any tests?

erights replied:
Is it ok that, once created, it lasts and remains active forever? If not, how about returning a kit of a notifier and a cancel object?

erights commented on packages/SwingSet/src/vats/vat-timerWrapper.js:
```diff
@@ -41,6 +42,29 @@ export function buildRootObject(vatPowers) {
+        );
+
+        const index = D(timerNode).createRepeater(delaySecs, interval);
+        console.log(`TIMER: index: ${index}`);
```
Are these `console.log` calls in the production code? (also line 60 below) Or are they debugging vestiges? I notice that the other console methods have no such `console` calls.

katelynsills replied:
I think we should remove the `console.log`s

erights commented on packages/SwingSet/src/vats/vat-timerWrapper.js:
```diff
@@ -41,6 +42,29 @@ export function buildRootObject(vatPowers) {
+        console.log(`TIMER: index: ${index}`);
+        const { notifier, updater } = makeNotifierKit();
+        const updateHandler = harden({
+          wake: timestamp => {
+            console.log(`TimerNotifier tick ${timestamp}`);
+            updater.updateState(timestamp);
+          },
```
Assuming that the `console` call is a vestige (see previous comment) how about
```suggestion
          wake: updater.updateState,
```

katelynsills commented on packages/SwingSet/src/vats/vat-timerWrapper.js:
```diff
@@ -41,6 +42,29 @@ export function buildRootObject(vatPowers) {
       },
+      createNotifier(delaySecs, interval) {
+        assert(
+          Nat(delaySecs) >= 0,
```
Nat() already errors if it is negative, so this can be:
```suggestion
          Nat(delaySecs)
```

Chris-Hibbert replied:
@erights points out that the `assert` around `Nat(delaySecs) >= 0` can't ever be reached. I've changed them to bare `Nat(delaySecs)` next to `assert(Nat(interval) > 0, '... must be a positive integer');`

erights replied:
Doesn't look fixed. You got rid of the `>= 0`, but there's still a pointless `assert` with unreachable details around the `Nat` call. Did you forget to push?

katelynsills commented on packages/SwingSet/src/vats/vat-timerWrapper.js:
```diff
@@ -41,6 +42,29 @@ export function buildRootObject(vatPowers) {
         repeaters.set(index, vatRepeater);
         return vatRepeater;
       },
+      createNotifier(delaySecs, interval) {
```
Should this be called `createNotifier` or `makeNotifier`? I thought we generally wanted to use `make`.

erights replied:
Please rename both, but leave the old one too, as deprecated.

katelynsills reviewed (changes requested):
I agree with @erights' comments. 

There's a lot of technical debt in this file, in that it lacks tests and documentation. I don't think we have enough time to clean all of it up, but I think before this gets reviewed again, we should add at a minimum:

* Types with documenting comments for at least this method
* Unit tests for the method
* Integration tests such as in `test-home.js` in cosmic-swingset to make sure the functionality in this method is as expected from the user's perspective
* Matching changes to `manualTimer` such that a developer can use the same functionality in their tests

Chris-Hibbert commented:
Just to be clear: 
> I think before this gets reviewed again, we should add at a minimum:
> * Types with documenting comments for at least this method
Added `createNotifier()` to the type declaration that already existed in `zoe/tools/types.js`.

> * Unit tests for the method
We don't have unit tests for vat code in SwingSet. I explored adding them, and didn't find a way.

> * Integration tests such as in test-home.js in cosmic-swingset to make sure the functionality in this method is as expected from the user's perspective
I added tests for `createNotifier()` and `createRepeater()` there.

> * Matching changes to manualTimer such that a developer can use the same functionality in their tests
Done

Also: 

 * I left the methods named `createNotifier()` and `createRepeater()` and explained why. 

 * Nat(0) doesn't fail, but it returns a falsy value. The asserts for non-negative and positive `Nats` were left as `assert(Nat(var) > 0, ...)` and `assert(Nat(var) >= 0, ...)` so they look parallel. 

And @erights suggested returning a canceller for the notifier. I think allowing the creator of the loan to break the notifier would have bad consequences, so I didn't add this.

erights commented on packages/SwingSet/src/vats/vat-timerWrapper.js:
```diff
@@ -20,11 +22,11 @@ export function buildRootObject(vatPowers) {
       createRepeater(delaySecs, interval) {
         assert(
           Nat(delaySecs) >= 0,
-          details`createRepeater's first parameter must be a non-negative integer. ${delaySecs}`,
+          details`createRepeater's first parameter must be a non-negative integer: ${delaySecs}`,
         );
```
This assert can never fail. Rather, of `delaySecs` is not a Nat, `Nat(delaySecs)` will throw and `assert` will never be called.

erights replied:
Doesn't look fixed. Did you forget to push?

katelynsills commented on packages/cosmic-swingset/test/test-home.js:
```diff
@@ -102,6 +102,46 @@ test.serial('home.wallet - receive zoe invite', async t => {
+  t.is(update1.updateCount, 2);
+  const update2 = await E(notifier).getUpdateSince(update1.updateCount);
+  t.is(update2.updateCount, 3);
+  t.truthy(update2.value > update1.value);
```
Can we add a comment about the actual value being a timestamp, and therefore we can't test against it directly, only relatively?

katelynsills reviewed (changes requested):
Thanks for adding the tests and types! 

For the types, I think as written, they are not working. In order for them to work properly, we need to move the type definition from Zoe to Swingset and change the references accordingly. 

I agree with @erights that we should rename the functions to `makeNotifier` and `makeRepeater` and deprecate `createRepeater`.

If we did want to add unit tests, I think it would be as easy as importing `buildRootObject` from `vat-timerWrapper` and mocking `vatPowers` as necessary. Alternatively, one thing that we did in cosmic-swingset with vat code was to create a separate file with the prefix `lib-` such as `lib-board` or `lib-wallet` that actually had the code which could be tested separately. Then that code was imported in the `vat-` file.

Chris-Hibbert reviewed (commented):
> If we did want to add unit tests, I think it would be as easy as importing buildRootObject from vat-timerWrapper and mocking vatPowers as necessary. Alternatively, one thing that we did in cosmic-swingset with vat code was to create a separate file with the prefix lib- such as lib-board or lib-wallet that actually had the code which could be tested separately. Then that code was imported in the vat- file.

I don't think mocking vatPowers is simple enough to be worth the trouble. The piece of vatPowers used in this vat defines the `D()` operator, which provides access to device nodes. So we'd also need to mock the device affordances that the vat relies on. 

The logic in this vat is pretty trivial, and is easy enough to exercise in integration tests.

erights commented on packages/zoe/src/contracts/priceAggregator.js:
```diff
@@ -9,8 +9,15 @@ import {
 
-import '../../exported';
-import './exported';
+// the bulk import style for typescript declarations doesn't seem to work here.
```
@michaelfig any idea why the bulk import style does not work here?

katelynsills commented on packages/zoe/tools/manualTimer.js:
```diff
@@ -72,38 +106,22 @@ export default function buildManualTimer(log, startValue = 0) {
       return harden(baseTimes);
     },
     createRepeater(delay, interval) {
```
should this be `delaySecs` to match the old, actual implementation?

Chris-Hibbert replied:
No, it should be `delay` to match the old, deprecated typescript declaration.

Chris-Hibbert commented:
> @michaelfig any idea why the bulk import style does not work here?

@katelynsills showed me how to fix it.

katelynsills commented on packages/zoe/src/contracts/priceAggregator.js:
```diff
@@ -9,9 +9,6 @@ import {
   makeOnewayPriceAuthorityKit,
 } from '../contractSupport';
 
-import '../../exported';
-import './exported';
-
```
I think this should be: `import '../../exported';`

Chris-Hibbert replied:
`yarn lint` and vsCode are happy without it.

katelynsills commented on packages/zoe/tools/priceAuthorityRegistry.js:
```diff
@@ -4,8 +4,6 @@ import { E } from '@agoric/eventual-send';
 import { makeStore } from '@agoric/store';
 import { assert, details } from '@agoric/assert';
 
-import '../exported';
-
```
I think this should remain

Chris-Hibbert replied:
`yarn lint` and vsCode are happy without it. I'm not going to try to explain that.

katelynsills reviewed (approved):
Looks good to me! Approval conditional on the types all working, and a slight nitpick of `delay` vs `delaySecs` straightened out.

Chris-Hibbert commented on packages/SwingSet/src/vats/vat-timerWrapper.js:
```diff
@@ -17,17 +19,19 @@ export function buildRootObject(vatPowers) {
-        );
+      // deprecated in favor of makeRepeater().
+      // TODO(#2164): remove before Beta
+      createRepeater(delay, interval) {
```
Probably better that way. The typescript declaration was `delay`, so I thought I was reverting for consistency, but there's no value in changing it here where it was inconsistent with the docs.
